### PR TITLE
Don't merge yet: Create accounts automatically when authenticating from HTTP header

### PR DESCRIPTION
**Pull request by selfhoster1312**
I tried to setup SSO authentication via HTTP header on the [Yunohost selfhosting distro package](https://github.com/YunoHost-Apps/navidrome_ynh/pull/85) but ran into the "User passed in header not found error" [described here](https://github.com/navidrome/navidrome/pull/1152#issuecomment-939919319).

This PR simply creates a new account when the user account was not found, when the HTTP reverse proxy is setup to be trusted for authentication. Putting an empty password creates runtime errors, so i used a random password.

The problem is now the user can't change the password to use in the Subsonic API. What should we do ?

- disable current password field entirely when HTTP auth is enabled?
- prefill it?
- something else?

In our specific case, we can access the plaintext user password via HTTP headers so we could inject it like that. However that is not true of most SSO systems, and it's always better if we can setup a different token/password for API access, while having the SSO system handle user passwords directly.

### CODE REVIEW

**Comment by deluan** [comment]
Thanks for the PR. The long term solution for this is what you mentioned: Make the API password independent from the user's password. I'm thinking of working on this after next release (end of this month). Until then, I don't think adding this is a good idea.

**Comment by selfhoster1312** [comment]
Thanks! Are you ok with temporary solution to disable password check when changing password when header auth is enabled? That makes both SSO functional, and support for 3rd party clients using the API (by changing password just putting nothing or wrong value in current password field).

It's a hack and i hope to have better support in the future, but if it can enable good SSO support in the next minor release that will be :1st_place_medal: 

I added a commit to do this.

**Comment by selfhoster1312** [comment]
I just did. FYI i had to single quote the *.go pattern to make it work, like this: `goimports -w $(find . -name '*.go' | grep -v _gen.go) && go mod tidy`.

**Comment by deluan** [comment]
Ops, you forgot to update/add tests

**Comment by selfhoster1312** [comment]
I broke a test because no new/old password means password change is validated (i thought this function not called in that case), and others failed for trying access to config's header auth settings which was not set in test suite. Should be ok now.

File: persistence/user_repository.go [hunk 1]
```diff
@@ -206,12 +206,15 @@ func validatePasswordChange(newUser *model.User, logged *model.User) error {
...
+	if newUser.NewPassword == "" {
+		if newUser.CurrentPassword == "" {
+			return nil
+		}
+		err.Errors["password"] = "ra.validation.required"
 	}
-	if newUser.CurrentPassword != "" {
-		if newUser.NewPassword == "" {
-			err.Errors["password"] = "ra.validation.required"
+	if conf.Server.ReverseProxyUserHeader == "" {
```
Comment (deluan) [inline persistence/user_repository.go:215]: The default for this option is not empty, so with this change it means that by default, this validation is completely skipped.

File: persistence/user_repository.go [hunk 2]
```diff
@@ -206,12 +206,15 @@ func validatePasswordChange(newUser *model.User, logged *model.User) error {
...
+		if newUser.CurrentPassword == "" {
+			err.Errors["currentPassword"] = "ra.validation.required"
 		}
 		if newUser.CurrentPassword != logged.Password {
```
Comment (deluan) [inline persistence/user_repository.go:219]: IMO, a better solution is to create the user with an empty password (only via SSO, ofc), and check here if` logged.Password` is empty. If it is, we don't need to do this check.

**Review by deluan — CHANGES_REQUESTED** [review]
Hey, check my comments, and let me know if you agree/need clarification. Also let me know if you will be able to work on this this week, or else I can try to finish it for next release (end of next week)

**Comment by selfhoster1312** [comment]
Very good points! I didn't use empty password because it caused other problems in parts of the code (not sure where anymore), but i prefixed the passwords as you suggested. Does that look better?

**Review by deluan — APPROVED** [review]

**Pull request merged**